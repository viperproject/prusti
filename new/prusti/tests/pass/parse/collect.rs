// compile-flags: -Zprint-desugared-specs -Zprint-collected-verification-items -Zhide-uuids
// normalize-stdout-test: "[a-z0-9]{32}" -> "$(NUM_UUID)"
// normalize-stdout-test: "[a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12}" -> "$(UUID)"
// ignore-test

#![feature(register_tool)]
#![register_tool(prusti)]

use prusti_contracts::*;

#[requires(true)]
fn test1() {}

#[ensures(true)]
fn test2() {}

fn test3() {
    for _ in 0..2 {
        invariant!(true)
    }
}

#[requires(true)]
#[ensures(true)]
fn test4() {
    for _ in 0..2 {
        invariant!(true)
    }
}

fn main() {}
