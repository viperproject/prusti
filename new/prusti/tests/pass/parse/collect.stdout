// compile-flags: -Zprint-desugared-specs -Zprint-collected-verification-items -Zhide-uuids
// normalize-stdout-test: "[a-z0-9]{32}" -> "$(NUM_UUID)"
// normalize-stdout-test: "[a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12}" -> "$(UUID)"

#![feature(register_tool)]
#![register_tool(prusti)]
#[prelude_import]
use std::prelude::v1::*;
#[macro_use]
extern crate std;

use prusti_contracts::*;

#[prusti::spec_only]
#[prusti::spec_id = "$(NUM_UUID)"]
#[prusti::assertion =
  "{/"kind/":{/"Expr/":{/"spec_id/":/"$(UUID)/",/"expr_id/":101}}}"]
fn prusti_pre_item_test1_$(NUM_UUID)() {

    #[prusti::expr_id = "$(NUM_UUID)_101"]
    || -> bool { true };
}
#[prusti::pre_spec_id_ref = "$(NUM_UUID)"]
fn test1() { }

#[prusti::spec_only]
#[prusti::spec_id = "$(NUM_UUID)"]
#[prusti::assertion =
  "{/"kind/":{/"Expr/":{/"spec_id/":/"$(UUID)/",/"expr_id/":101}}}"]
fn prusti_post_item_test2_$(NUM_UUID)() {

    #[prusti::expr_id = "$(NUM_UUID)_101"]
    || -> bool { true };
}
#[prusti::post_spec_id_ref = "$(NUM_UUID)"]
fn test2() { }

fn test3() {
    for _ in 0..2 {


        {
            #[allow(dead_code)]
            enum ProcMacroHack { Value = ("true", 0).1, }
            macro_rules! proc_macro_call {
                () =>
                {
                    if false
                    {
                        # [prusti :: spec_only] #
                        [prusti :: spec_id =
                         "$(NUM_UUID)"] #
                        [prusti :: assertion =
                         "{/"kind/":{/"Expr/":{/"spec_id/":/"$(UUID)/",/"expr_id/":101}}}"]
                        let _prusti_loop_invariant =
                        {
                            #
                            [prusti :: expr_id =
                             "$(NUM_UUID)_101"] || ->
                            bool { true } ;
                        } ;
                    }
                }
            }
            if false {
                #[prusti::spec_only]
                #[prusti::spec_id = "$(NUM_UUID)"]
                #[prusti::assertion =
                  "{/"kind/":{/"Expr/":{/"spec_id/":/"$(UUID)/",/"expr_id/":101}}}"]
                let _prusti_loop_invariant =
                    {

                        #[prusti::expr_id =
                          "$(NUM_UUID)_101"]
                        || -> bool { true };
                    };
            }
        }
    }
}
#[prusti::spec_only]
#[prusti::spec_id = "$(NUM_UUID)"]
#[prusti::assertion =
  "{/"kind/":{/"Expr/":{/"spec_id/":/"$(UUID)/",/"expr_id/":101}}}"]
fn prusti_pre_item_test4_$(NUM_UUID)() {

    #[prusti::expr_id = "$(NUM_UUID)_101"]
    || -> bool { true };
}
#[prusti::spec_only]
#[prusti::spec_id = "$(NUM_UUID)"]
#[prusti::assertion =
  "{/"kind/":{/"Expr/":{/"spec_id/":/"$(UUID)/",/"expr_id/":101}}}"]
fn prusti_post_item_test4_$(NUM_UUID)() {

    #[prusti::expr_id = "$(NUM_UUID)_101"]
    || -> bool { true };
}
#[prusti::post_spec_id_ref = "$(NUM_UUID)"]
#[prusti::pre_spec_id_ref = "$(NUM_UUID)"]
fn test4() {
    for _ in 0..2 {
        {
            #[allow(dead_code)]
            enum ProcMacroHack { Value = ("true", 0).1, }
            macro_rules! proc_macro_call {
                () =>
                {
                    if false
                    {
                        # [prusti :: spec_only] #
                        [prusti :: spec_id =
                         "$(NUM_UUID)"] #
                        [prusti :: assertion =
                         "{/"kind/":{/"Expr/":{/"spec_id/":/"$(UUID)/",/"expr_id/":101}}}"]
                        let _prusti_loop_invariant =
                        {
                            #
                            [prusti :: expr_id =
                             "$(NUM_UUID)_101"] || ->
                            bool { true } ;
                        } ;
                    }
                }
            }
            if false {
                #[prusti::spec_only]
                #[prusti::spec_id = "$(NUM_UUID)"]
                #[prusti::assertion =
                  "{/"kind/":{/"Expr/":{/"spec_id/":/"$(UUID)/",/"expr_id/":101}}}"]
                let _prusti_loop_invariant =
                    {

                        #[prusti::expr_id =
                          "$(NUM_UUID)_101"]
                        || -> bool { true };
                    };
            }
        }
    }
}
fn main() { }
Collected verification items 5:
procedure: collect::test3[0] at $DIR/collect.rs:16:1: 20:2
procedure: collect::main[0] at $DIR/collect.rs:30:1: 30:13
procedure: collect::test1[0] at $DIR/collect.rs:11:1: 11:14
procedure: collect::test2[0] at $DIR/collect.rs:14:1: 14:14
procedure: collect::test4[0] at $DIR/collect.rs:1:1: 23:17
